app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -0,0 +1,182 @@
+import { ethErrors } from 'eth-json-rpc-errors'
+import validUrl from 'valid-url'
+// import { NETWORK_TO_NAME_MAP as DEFAULT_NETWORK_MAP } from '../../../controllers/network/enums'
+import { omit } from 'lodash'
+import { isPrefixedFormattedHexString } from '../../util'
+import { MESSAGE_TYPE } from '../../../../../shared/constants/app'
+
+const addEthereumChain = {
+  methodNames: [MESSAGE_TYPE.ADD_ETHEREUM_CHAIN],
+  implementation: addEthereumChainHandler,
+}
+export default addEthereumChain
+
+async function addEthereumChainHandler(
+  req,
+  res,
+  _next,
+  end,
+  { addCustomRpc, customRpcExistsWith, requestUserApproval },
+) {
+  if (!req.params?.[0] || typeof req.params[0] !== 'object') {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected single, object parameter. Received:\n${req.params}`,
+      }),
+    )
+  }
+
+  const { origin } = req
+
+  const {
+    chainId,
+    chainName = null,
+    blockExplorerUrls = null,
+    nativeCurrency = null,
+    rpcUrls,
+  } = req.params[0]
+
+  const otherKeys = Object.keys(
+    omit(req.params[0], [
+      'chainId',
+      'chainName',
+      'blockExplorerUrls',
+      'iconUrl',
+      'rpcUrls',
+      'nativeCurrency',
+    ]),
+  )
+
+  if (otherKeys.length > 0) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Received unexpected keys on object parameter. Unsupported keys:\n${otherKeys}`,
+      }),
+    )
+  }
+
+  const firstValidRPCUrl = Array.isArray(rpcUrls)
+    ? rpcUrls.find((rpcUrl) => validUrl.isHttpsUri(rpcUrl))
+    : null
+
+  const firstValidBlockExplorerUrl =
+    blockExplorerUrls !== null && Array.isArray(blockExplorerUrls)
+      ? blockExplorerUrls.find((blockExplorerUrl) =>
+          validUrl.isHttpsUri(blockExplorerUrl),
+        )
+      : null
+
+  if (!firstValidRPCUrl) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected an array with at least one valid string HTTPS url 'rpcUrls', Received:\n${rpcUrls}`,
+      }),
+    )
+  }
+
+  const _chainId = typeof chainId === 'string' && chainId.toLowerCase()
+  if (!isPrefixedFormattedHexString(_chainId)) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected 0x-prefixed, unpadded, non-zero hexadecimal string 'chainId'. Received:\n${chainId}`,
+      }),
+    )
+  }
+
```

Comment:
Here, we should now call `isSafeChainId` as done in #10224.

Reply:
done